### ISSUE TITLE
Unhelpful log prefix on packets dropped by --expect-npc

### ISSUE THREAD

**Issue by awh** [weaveworks/weave#2546]
New message TBD.

**Comment by awh** [comment]
https://weaveworks.slack.com/archives/kubernetes/p1476263269002516

**Comment by rade** [comment]
- "WEAVE-NPC dropped packet" (but why?)
- "WEAVE-NPC dropped packet due to policy" ("packet" is a bit obscure)
- "WEAVE-NPC denied/rejected/etc connection due to policy" (but the cause may be there actually is *no* policy)
- ?

**Comment by awh** [comment]
Thinking about it further, this is probably rendered moot by weaveworks/weave-npc#7

**Comment by rade** [comment]
I don't think so; not everybody is running prom.

**Comment by awh** [comment]
My thinking is that once weave-npc is consuming the dropped packets via NFLOG, they will no longer be logged as they are now with --log-prefix; we'll have complete freedom in what we display in the weave-npc logs (in addition to exposing them as Prom metrics)

**Comment by rade** [comment]
I see. Still, IMO improving the message here is a quick win.

**Comment by marccarre** [comment]
What would we [the users] do in case of a connection we [they] genuinely want to drop?
Do we expect it to be done by the NPC or is it okay (or even expected) to continue processing iptables rules?

What about something like the below?

1. `WEAVE-NPC denied connection. Please ensure policies are configured at NPC-level.`
2. `WEAVE-NPC denied connection. If unexpected, please ensure policies are configured at NPC-level.`
3. `WEAVE-NPC denied connection as a default behavior.`
4. `WEAVE-NPC denied connection as a default behavior. Please ensure policies are configured at NPC-level.`
5. `WEAVE-NPC denied connection as a default behavior. If unexpected, please ensure policies are configured at NPC-level.`

**Comment by awh** [comment]
> What would we [the users] do in case of a connection we [they] genuinely want to drop?
> Do we expect it to be done by the NPC or is it okay (or even expected) to continue processing iptables rules?

If a packet reaches the LOG rule, it means that:

* the user has taken positive action to enable the `DefaultDeny` policy (it's off by default, in which case the NPC implements `DefaultAllow` and the packet never makes it to this point) and
* they didn't configure an explicit network policy allowing the traffic.

Consequently, if a packet reaches the LOG/DROP rules it means either that:

* something unauthorised is attempting access, or
* they have misconfigured their network policies and something which should be allowed access is being blocked.

In both of these case, we need to alert the user so they can take action. In other words, there are no cases in which we want to silently drop traffic.

Finally, some more context - the  `-j LOG --log-prefix=WEAVE-NPC:` rule results in messages like this in the kernel log:

```
[14927.710232] WEAVE-NPC:=IN= OUT=eth0 SRC=192.168.2.221 DST=93.184.216.34 LEN=84 TOS=0x00 PREC=0x00 TTL=64 ID=20511 DF PROTO=ICMP TYPE=8 CODE=0 ID=21213 SEQ=1
```

The 'WEAVE-NPC:' string is the only thing we get to control here. Personally I think something like this is enough:

```
[14927.710232] Connection blocked by Weave NPC: =IN= OUT=eth0 SRC=192.168.2.221 DST=93.184.216.34 LEN=84 TOS=0x00 PREC=0x00 TTL=64 ID=20511 DF PROTO=ICMP TYPE=8 CODE=0 ID=21213 SEQ=1
```

Because I don't think a log statement is an appropriate place to list what, all possible reasons why, and the possible remedies (imagine there can be hundreds of instances of this log line due to retries). I'm not sure @rade agrees with me though!

**Comment by rade** [comment]
We could make sure that googling for "Connection blocked by Weave NPC" gets you to a docs page on our web site that has more explanation.

**Comment by marccarre** [comment]
Yes, I also was reluctant to make the log messages too long.
Maybe a combination of:

* a short message in the logs like what you suggested @awh, 
* a well referenced documentation if someone googles for "Connection blocked by Weave NPC" like what you suggested @rade, but also
* a link to the documentation (e.g. a very short bit.ly link) in the log message itself,

would be the best solution then?

**Comment by marccarre** [comment]
---
title: Weave NPC
menu_order: 70
---

* [> "Connection dropped by Weave NPC"](#connection-dropped)

##<a name="connection-dropped"></a>"Connection dropped by Weave NPC"

When using Weave together with [Kubernetes](http://kubernetes.io) and [Weave NPC](https://github.com/weaveworks/weave-npc), Weave's Kubernetes Network Policy Controller, `weave launch --expect-npc` will set up the two below default `iptables` rules, in case none of the NPC rules apply:

    LOG --log-prefix="Connection dropped by Weave NPC (see: bit.ly/2eNRyDF)"
    DROP

which can eventually generate kernel log messages like the below one:

    [12345.678901] Connection dropped by Weave NPC (see: bit.ly/2eNRyDF):=IN= OUT=eth0 SRC=xxx.xxx.xxx.xxx DST=yyy.yyy.yyy.yyy LEN=xx TOS=0x00 PREC=0x00 TTL=64 ID=1001 DF PROTO=ICMP TYPE=8 CODE=0 ID=1337 SEQ=1

If you ever see the above message and did not expect it, then:

* you must have enabled the `DefaultDeny` policy in NPC -- which is off by default, in which case NPC implements `DefaultAllow` and the packet never makes it to the above default rules as it is accepted by NPC much earlier;
* you did not configure a network policy allowing the traffic for this packet.

This could mean that either:

* you have misconfigured your network policies and something which should be allowed access is being blocked, in which case you may want to review your network policies, or
* something genuinely unauthorized is attempting access, in which case you may want to take action.

**Comment by marccarre** [comment]
@rade / @awh, please let me know if you are happy with the above / #2566.
Hopefully this will help the user to diagnose their setup and point them to the right direction?

**Comment by brb** [comment]
Just as an example, ufw logs dropped packets in the following format:
```
[38518.786179] [UFW BLOCK] IN=wlp4s0 OUT= MAC=01:00:5e:00:00:01:58:98:35:22:b0:4c:08:00 SRC=192.168.1.254 DST=224.0.0.1 LEN=36 TOS=0x00 PREC=0xC0 TTL=1 ID=41080 DF PROTO=2
```

Re the URL: IMO, it's redundant, as a user is aware of the weave-npc presence.

**Comment by marccarre** [comment]
After discussion with @brb, @awh and @bboreham, it has been decided it was better to remove the link. I'll update #2566 accordingly.

**Comment by marccarre** [comment]
@rade, after discussion with @brb, @awh and @bboreham, it has been decided it was better to remove the link.

* #2566 has a log message with a short URL to where the documentation will eventually appear.
* #2573 has the same log message, without the short URL.

**Comment by bboreham** [comment]
Fixed by #2573
